### PR TITLE
test: Add PT Executor Test

### PR DESCRIPTION
**Pull request by retrage**
## Type of PR

- [ ] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [x] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [ ] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Related Issue

#12

## Importance of PR

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [x] Medium (requires several hours to a day of review)
    - [ ] Small (trivial changes, such as build error)

## PR Overview

This PR adds a test for PT Executor as planned in #12.

## Concerns (Optional)

- [ ] Performance
- [x] Source Code Quality

This test expects the patched `tee` ELF is located in the fixed location (`/usr/bin/tee.patched`) for simplicity. Ideally, the
test script should automatically build the glibc and patch the target PUT. This should be fixed later. The purpose of this PR is to reproduce the test results of my local test environment for other people.

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [x] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [x] The PR title is a summary of the changes.
- [x] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

### CODE REVIEW

**Comment by retrage** [comment]
The test results on my local environment:

```
$ ctest -R pt_executor.run -V
UpdateCTestConfiguration  from :/home/retrage/src/fuzzuf/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/retrage/src/fuzzuf/build/DartConfiguration.tcl
Test project /home/retrage/src/fuzzuf/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 25
    Start 25: pt_executor.run

25: Test command: /usr/bin/sudo "-E" "/home/retrage/src/fuzzuf/build/test/executor/pt_executor/test-pt-executor-run"
25: Test timeout computed to be: 10000000
25: Running 1 test case...
25: [*] output_file_path = /tmp/fuzzuf_test.z5hUj6/output/result
25: cargv[0]: /home/retrage/src/AFLplusplus/pt_mode/pt_proxy/pt-proxy-fast
25: Run: 
25: /home/retrage/src/AFLplusplus/pt_mode/pt_proxy/pt-proxy-fast 
25: /home/retrage/src/AFLplusplus/pt_mode/pt_proxy/pt-proxy-fast 
25: -- 
25: /usr/bin/tee.patched 
25: /tmp/fuzzuf_test.z5hUj6/output/result 
25: (null) 
25: 
25: 
25: Exec Status 0 (pid 0)
25: 
25: 
25: *** No errors detected
25: 
1/1 Test #25: pt_executor.run ..................   Passed    0.32 sec

The following tests passed:
	pt_executor.run

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.32 sec
```

**Comment by retrage** [comment]
run tests

File: test/executor/pt_executor/run.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  // PUTExitReasonType::FAULT_NONE.
+  BOOST_CHECK_EQUAL(executor.GetExitStatusFeedback().exit_reason,
+                    PUTExitReasonType::FAULT_NONE);
+
+  // (2) Check if The target PUT received the input value.
+  auto stdout_buffer = executor.MoveStdOut();
+  std::vector<std::uint8_t> expected_stdout{'a', 'b', 'c', 'd'};
+  BOOST_CHECK_EQUAL_COLLECTIONS(stdout_buffer.begin(), stdout_buffer.end(),
+                                expected_stdout.begin(), expected_stdout.end());
+
```
Comment (K-atc) [inline test/executor/pt_executor/run.cpp:92]: I can't grasp what PT mode feedback would be correct if it were included. Won't you at least make sure that "checking if something might correct exists"（"それっぽいものが入っているのか確認"）?

**Review by K-atc — CHANGES_REQUESTED** [review]
@retrage 
I assume this PR is a simple change. This initial review was done in a simplified manner.

File: test/executor/pt_executor/run.cpp (hunk 1 above)
Comment (retrage) [reply test/executor/pt_executor/run.cpp:92]: Thanks for the review. It's possible by checking `CountNonZeroBytes()` for the bitmaps.

Comment (K-atc) [reply test/executor/pt_executor/run.cpp:92]: :yosasou: go go!

Comment (retrage) [reply test/executor/pt_executor/run.cpp:92]: I added the coverage bitmap checks.

**Comment by retrage** [comment]
This PR should be merged after #59 is merged. I checked this test with https://github.com/fuzzuf/pt_mode/tree/213d6983efc51e70689048451eb40b246b07cba9.
Here is the test results in my local environment:

```
$ ctest -R pt_executor.run -V
UpdateCTestConfiguration  from :/home/retrage/src/fuzzuf/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/retrage/src/fuzzuf/build/DartConfiguration.tcl
Test project /home/retrage/src/fuzzuf/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 29
    Start 29: pt_executor.run

29: Test command: /usr/bin/sudo "-E" "/home/retrage/src/fuzzuf/build/test/executor/pt_executor/test-pt-executor-run"
29: Test timeout computed to be: 10000000
[sudo] password for retrage: 
29: Running 1 test case...
29: [*] output_file_path = /tmp/fuzzuf_test.Ioa9m3/output/result
29: cargv[0]: /home/retrage/src/AFLplusplus/pt_mode/pt_proxy/pt-proxy-fast
29: Run: 
29: /home/retrage/src/AFLplusplus/pt_mode/pt_proxy/pt-proxy-fast 
29: /home/retrage/src/AFLplusplus/pt_mode/pt_proxy/pt-proxy-fast 
29: -- 
29: /usr/bin/tee.patched 
29: /tmp/fuzzuf_test.Ioa9m3/output/result 
29: (null) 
29: 
29: 
29: Exec Status 0 (pid 0)
29: 
29: 
29: *** No errors detected
29: 
1/1 Test #29: pt_executor.run ..................   Passed    3.53 sec

The following tests passed:
        pt_executor.run

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   3.53 sec
```

**Comment by retrage** [comment]
run tests

**Review by K-atc — APPROVED** [review]
This PR introduces test `pt_executor.run`. @retrage tested it *locally*, and it was performed in successful.
The test now checks if target PUT returns non-empty coverage bitmap in step (3) (that was suggested by me).
This PR can be merged, I think.
お疲れ様でした～

**Pull request merged**